### PR TITLE
[GStreamer][MSE] Refactor demuxer handling in AppendPipeline

### PR DESCRIPTION
**Pull request by philn**
#### c75026b5ef156d070d279c9dc209d1bcb0f3c0c5
<pre>
[GStreamer][MSE] Refactor demuxer handling in AppendPipeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=248742">https://bugs.webkit.org/show_bug.cgi?id=248742</a>

Reviewed by Alicia Boya Garcia.

Instead of a boolean `hasDemuxer` we can check the element metadata for the `Demuxer` classifier.

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::AppendPipeline):

Canonical link: <a href="https://commits.webkit.org/258044@main">https://commits.webkit.org/258044@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/555716f4f7bf2f5f18d947c043100a96d60389cb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9892 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33789 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110047 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170321 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10831 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/458 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93148 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107893 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106530 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8178 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34787 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22822 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77758 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3586 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24346 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3610 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9721 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43845 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5523 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5387 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Review by ntrrgc — CHANGES_REQUESTED** [review]
There are two changes mixed here:

### 1. Replacing the `hasDemuxer` variable with a metadata check.

It's more complex, but okay.

### 2. Removing the `identity` element for `m_typefind`.

Yes, you could do that, yes you would remove three lines of code explicitly constructing an `identity` that way; but the idea is to aim for orthogonality.

Without the identity, the seeming simplicity you gain by removing those 3 lines you gain back in pipeline manipulations when adding the typefind to the pipeline.

I am not bothered by holding a reference to m_typefind in the class either. It's an element of the AppendPipeline like any other.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp [hunk 1]
```diff
@@ -176,7 +177,7 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
         GRefPtr<GstPad> identitySrcPad = adoptGRef(gst_element_get_static_pad(m_demux.get(), "src"));
         gst_pad_add_probe(identitySrcPad.get(), GST_PAD_PROBE_TYPE_BUFFER, reinterpret_cast<GstPadProbeCallback>(
             +[](GstPad *pad, GstPadProbeInfo*, AppendPipeline* appendPipeline) {
-                GRefPtr<GstCaps> caps = gst_pad_get_current_caps(pad);
+                auto caps = adoptGRef(gst_pad_get_current_caps(pad));
```
Comment (ntrrgc) [inline Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: Unnecessary unrelated change. In this case it's okay either way; but in general being explicit about types is not a bad thing, and just because we can set it to `auto` doesn't mean we need to or should every time.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp [hunk 2]
```diff
@@ -162,11 +161,13 @@ AppendPipeline::AppendPipeline(SourceBufferPrivateGStreamer& sourceBufferPrivate
     m_demuxerDataEnteringPadProbeInformation.probeId = gst_pad_add_probe(demuxerPad.get(), GST_PAD_PROBE_TYPE_BUFFER, reinterpret_cast<GstPadProbeCallback>(appendPipelinePadProbeDebugInformation), &m_demuxerDataEnteringPadProbeInformation, nullptr);
 #endif
 
-    if (hasDemuxer) {
+    auto elementClass = makeString(gst_element_get_metadata(m_demux.get(), GST_ELEMENT_METADATA_KLASS));
+    auto classifiers = elementClass.split('/');
+    if (classifiers.contains("Demuxer"_s)) {
         // These signals won't outlive the lifetime of `this`.
-        g_signal_connect(m_demux.get(), "no-more-pads", G_CALLBACK(+[](GstElement*, AppendPipeline* appendPipeline) {
+        g_signal_connect_swapped(m_demux.get(), "no-more-pads", G_CALLBACK(+[](AppendPipeline* appendPipeline) {
```
Comment (ntrrgc) [inline Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: This line is a good example of a very dangerous kind of "cleanup".

Cons:
* Undefined behavior in C++. [1]
* Breaks on platforms where the ABI specifies the callee cleans the stack.
* Breaks on platforms where function arguments are passed right-to-left.

Pros:
* It saves one argument for the callback.
* Makes the line 5 characters shorter.

[1] https://stackoverflow.com/a/2118983/1777162

**Review by ntrrgc — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 1 above)
Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: It's a leak fix though. Do you want this in a separate PR?

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 2 above)
Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: Are you saying we shouldn't use `g_signal_connect_swapped()`? Because it's currently used in many many places :)

Comment (ntrrgc) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: Then today we learn we have undefined behavior in many many places.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 1 above)
Comment (ntrrgc) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: My point was rather that there is no point on changing that.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 2 above)
Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: CC @carlosgcampos

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 1 above)
Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: No point in fixing a leak?

Comment (ntrrgc) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: Ah.

AH.

AAAH.

Okay, nvm. I hadn't noticed that. That should be in a different PR if for no other reason that otherwise it may get lost when doing backports.

Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: https://github.com/WebKit/WebKit/pull/7495

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 2 above)
Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: I don't think we've received any bug report about this. It may be undefined behavior but so far it has worked for us, unless proven otherwise :)

Comment (ntrrgc) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: It has worked for us because it just so happens that in x86_64 SysV calling convention, as well as in the calling conventions for ARM 32-bit and ARM 64-bit the first few arguments are passed as registers.

Comment (ntrrgc) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: After looking more into it (including looking at glib and libffi source code), it seems this is just how glib does things. It's indeed undefined behavior that works only thanks to specifics of popular ABIs, but it seems glib and GTK are okay with it. 

So no point in policing it in WebKit. It's still dangerous for any non-trivial set of arguments.

**Review by ntrrgc — APPROVED** [review]
r+ rebase and land

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 2 above)
Comment (ntrrgc) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:167]: Fun fact from today's talk in the GStreamer Conference 2023 about GStreamer in WebAssembly. This pattern breaks on Emscripten's ABI.
